tests/cpp/common/test_gpu_hist_util.cu

```diff
@@ -1,50 +1,62 @@
-#include "../../../src/common/device_helpers.cuh"
-#include "../../../src/common/hist_util.h"
-#include "gtest/gtest.h"
-#include "xgboost/c_api.h"
 #include <algorithm>
 #include <cmath>
+
+#include "gtest/gtest.h"
+
 #include <thrust/device_vector.h>
 #include <thrust/iterator/counting_iterator.h>
 
+#include "xgboost/c_api.h"
+
+#include "../../../src/common/device_helpers.cuh"
+#include "../../../src/common/hist_util.h"
+
+#include "../helpers.h"
+
 namespace xgboost {
 namespace common {
 
-void TestDeviceSketch(const GPUSet& devices) {
+void TestDeviceSketch(const GPUSet& devices, bool use_external_memory = false) {
   // create the data
   int nrows = 10001;
-  std::vector<float> test_data(nrows);
-  auto count_iter = thrust::make_counting_iterator(0);
-  // fill in reverse order
-  std::copy(count_iter, count_iter + nrows, test_data.rbegin());
-
-  // create the DMatrix
-  DMatrixHandle dmat_handle;
-  XGDMatrixCreateFromMat(test_data.data(), nrows, 1, -1,
-                         &dmat_handle);
-  auto dmat = static_cast<std::shared_ptr<xgboost::DMatrix> *>(dmat_handle);
-
-  // parameters for finding quantiles
+  std::shared_ptr<xgboost::DMatrix> *dmat = nullptr;
+
+  size_t num_cols = 1;
+  if (!use_external_memory) {
+     std::vector<float> test_data(nrows);
+     auto count_iter = thrust::make_counting_iterator(0);
+     // fill in reverse order
+     std::copy(count_iter, count_iter + nrows, test_data.rbegin());
+
+     // create the DMatrix
+     DMatrixHandle dmat_handle;
+     XGDMatrixCreateFromMat(test_data.data(), nrows, 1, -1,
+                            &dmat_handle);
+     dmat = static_cast<std::shared_ptr<xgboost::DMatrix> *>(dmat_handle);
+  } else {
+     auto sp_dmat = CreateSparsePageDMatrix(nrows * 3, 128UL); // 3 entries/row
+     dmat = new std::shared_ptr<xgboost::DMatrix>(std::move(sp_dmat));
+     num_cols = 5;
+  }
+
   tree::TrainParam p;
   p.max_bin = 20;
   p.gpu_id = 0;
   p.n_gpus = devices.Size();
-  // ensure that the exact quantiles are found
-  int gpu_batch_nrows = nrows * 10;
+  int gpu_batch_nrows = 0;
 
   // find quantiles on the CPU
   HistCutMatrix hmat_cpu;
   hmat_cpu.Init((*dmat).get(), p.max_bin);
 
   // find the cuts on the GPU
-  const SparsePage& batch = *(*dmat)->GetRowBatches().begin();
   HistCutMatrix hmat_gpu;
-  DeviceSketch(batch, (*dmat)->Info(), p, &hmat_gpu, gpu_batch_nrows);
+  (void)DeviceSketch(p, gpu_batch_nrows, (*dmat).get(), &hmat_gpu);
```

Comment:
No need to cast to `void`.

Reply:
this is to indicate that the return value is ignored. i can do it c++ style if you want!